File: call_wind.py
#!/usr/bin/env python3
"""
Wind AI — 课程知识库问答 API caller with SSE streaming support.
Usage: python3 call_wind.py "用户问题" [conversation_id]
"""

import sys
import json
import urllib.request
import urllib.error

API_URL = "https://api-spectra.duplik.cn/v1/conversations"
TOKEN = "sk-hQ9Izn7hfWgZsuyagHNZM85NC1XEH0KV"
AGENT_ID = "SP782317380904940"


def call_wind(message: str, conversation_id: str = "") -> dict:
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {TOKEN}"
    }

    body = json.dumps({
        "inputs": {},
        "query": message,
        "message": message,
        "conversation_id": conversation_id,
        "agent_id": AGENT_ID,
        "enable_websearch": False,
        "stream": True,
        "response_mode": "streaming"
    }).encode("utf-8")

    req = urllib.request.Request(API_URL, data=body, headers=headers, method="POST")

    full_text = ""
    returned_conversation_id = conversation_id
    buffer = ""

    try:
        with urllib.request.urlopen(req, timeout=60) as response:
            decoder_buffer = b""
            while True:
                chunk = response.read(1024)
                if not chunk:
                    break
                decoder_buffer += chunk
                try:
                    decoded = decoder_buffer.decode("utf-8")
                    decoder_buffer = b""
                except UnicodeDecodeError:
                    continue

                buffer += decoded
                lines = buffer.split("\n")
                buffer = lines.pop()  # keep incomplete line

                for line in lines:
                    trimmed = line.strip()
                    if not trimmed.startswith("data:"):
                        continue
                    json_str = trimmed[5:].strip()
                    if not json_str or json_str == "[DONE]":
                        continue
                    try:
                        data = json.loads(json_str)
                        event_type = data.get("type", "")
                        extra = data.get("extra", {})

                        if event_type == "stream" and data.get("message"):
                            full_text += data["message"]

                        if extra.get("conversation_id"):
                            returned_conversation_id = extra["conversation_id"]

                    except json.JSONDecodeError:
                        continue

    except urllib.error.HTTPError as e:
        error_body = e.read().decode("utf-8", errors="replace")
        print(json.dumps({"error": f"HTTP {e.code}: {error_body}"}))
        sys.exit(1)
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)

    return {
        "answer": full_text.strip(),
        "conversation_id": returned_conversation_id
    }


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: call_wind.py <message> [conversation_id]"}))
        sys.exit(1)

    user_message = sys.argv[1]
    conv_id = sys.argv[2] if len(sys.argv) > 2 else ""

    result = call_wind(user_message, conv_id)
    print(json.dumps(result, ensure_ascii=False))

File: call_wind.js
#!/usr/bin/env node
/**
 * Wind AI — Node.js 调用脚本（无需 Python，需 Node.js 18+ 或 WorkBuddy 内置 Node）
 * 用法: node call_wind.js "你的问题" [conversation_id]
 */

const API_URL = "https://api-spectra.duplik.cn/v1/conversations";
const TOKEN = "sk-hQ9Izn7hfWgZsuyagHNZM85NC1XEH0KV";
const AGENT_ID = "SP782317380904940";

const [, , message, conversationId = ""] = process.argv;

if (!message) {
  console.log(JSON.stringify({ error: "Usage: node call_wind.js <message> [conversation_id]" }));
  process.exit(1);
}

async function callWind(msg, convId) {
  const body = JSON.stringify({
    inputs: {},
    query: msg,
    message: msg,
    conversation_id: convId,
    agent_id: AGENT_ID,
    enable_websearch: false,
    stream: true,
    response_mode: "streaming",
  });

  const res = await fetch(API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${TOKEN}`,
    },
    body,
  });

  if (!res.ok) {
    const text = await res.text();
    console.log(JSON.stringify({ error: `HTTP ${res.status}: ${text}` }));
    process.exit(1);
  }

  let fullText = "";
  let returnedConvId = convId;
  const decoder = new TextDecoder();
  let buffer = "";

  for await (const chunk of res.body) {
    buffer += decoder.decode(chunk, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop(); // 保留不完整的行

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed.startsWith("data:")) continue;
      const jsonStr = trimmed.slice(5).trim();
      if (!jsonStr || jsonStr === "[DONE]") continue;

      try {
        const data = JSON.parse(jsonStr);
        if (data.type === "stream" && data.message) {
          fullText += data.message;
        }
        if (data.extra?.conversation_id) {
          returnedConvId = data.extra.conversation_id;
        }
      } catch {
        // 跳过解析失败的行
      }
    }
  }

  console.log(JSON.stringify({ answer: fullText.trim(), conversation_id: returnedConvId }, null, 0));
}

callWind(message, conversationId).catch((err) => {
  console.log(JSON.stringify({ error: String(err) }));
  process.exit(1);
});
